src/peer_connection.rs

```diff
@@ -27,40 +28,96 @@ impl Handshake {
 }
 
 pub struct PeerConnection {
-    stream: TcpStream,
+    // stream: TcpStream,
     bitfield: Vec<u8>,
+    outgoing_tx: mpsc::Sender<message::Message>,
 }
 
 pub struct PeerConnections {
     connections: Vec<PeerConnection>,
+    peer_incoming_rxs: Vec<mpsc::Receiver<message::Message>>,
 }
 
 impl PeerConnection {
-    pub fn new(stream: TcpStream) -> PeerConnection {
-        PeerConnection {
-            stream,
+    pub fn new(peer: &tracker::Peer,
+               metadata: &parser::TorrentMetadata) -> (PeerConnection, mpsc::Receiver<message::Message>) {
+        let (incoming_tx, incoming_rx) = mpsc::channel(1024);
+        let (outgoing_tx, outgoing_rx) = mpsc::channel(1024);
+
+        let socket_addr = SocketAddr::from(SocketAddrV4::new(peer.ip, peer.port));
+        tokio::spawn(peer_connection_background(socket_addr.clone(), metadata.clone(), incoming_tx, outgoing_rx));
+
+        (PeerConnection {
+            // stream,
             bitfield: Vec::new(),
+            outgoing_tx,
+        }, incoming_rx)
+    }
+
+}
+
+async fn peer_connection_background(socket_addr: SocketAddr,
+                                    metadata: parser::TorrentMetadata,
+                                    mut incoming_tx: mpsc::Sender<message::Message>,
+                                    mut outgoing_rx: mpsc::Receiver<message::Message>)
+    -> Result<(), String> {
+    let mut tcp_stream = perform_handshake(&socket_addr, &metadata).await.map_err(|e| e.to_string())?;
+
+    loop {
+        while let Some(outgoing_message) = outgoing_rx.try_recv().ok() {
```

Comment:
My understanding of this is that we can send anything from the outgoing_tx. We don't do anything with this until we call `recv()` (or `try_recv()`) on the outgoing_rx, which will return whatever is sent from the outgoing_tx. 

I can imagine we might process an incoming tcpstream message from incoming_rx and determine some response which the outgoing_rx sends, and then is received here.

I noticed playing around with mpsc we can't rely on the order that anything happens in. Probably won't impact us much but worth noting.